data_collection/timeprof_matrix_bot.py

```diff
@@ -483,11 +480,18 @@ def create_next_sample_time(self, prev_sample_time, rate):
         next_sample_time = prev_sample_time + timedelta(minutes=interval)
         return next_sample_time
 
-    def schedule_next_sample(self, user_id, sample_time):
+    async def schedule_next_sample(self, user_id, sample_time):
         loop = asyncio.get_event_loop()
+
+        rate = self.database.get_rate(user_id)
+        new_sample_time = self.create_next_sample_time(sample_time, rate)
+        await loop.create_task(self.run_at(sample_time, self.schedule_next_sample(user_id, new_sample_time)))
+        logging.info("Scheduled scheduling of next sample time at {}".format(sample_time))
+
         self.database.set_next_sample_time(user_id, sample_time)
```

Comment:
collect_user_activity calls get_next_sample_time. What happens if schedule_next_sample executes before collect_user_activity?

Could we just keep the scheduler_next_sample call in collect_user_activity but make sure messages are sent at the end. Or make sure messages are sent in their own tasks.

Reply:
Good catch.
I think it would be nice to keep responsibilities separate between the functions. I'll push an update!

Reply:
Now I'm starting to think about concurrent access in general (for this particular change it seems fine).